cdap-common/src/main/java/io/cdap/cdap/common/internal/remote/RemoteClient.java

```diff
@@ -155,14 +169,23 @@ public HttpURLConnection openConnection(String resource) throws IOException {
     urlConn.setDoInput(true);
     RemoteAuthenticator authenticator = getAuthenticator();
     if (authenticator != null) {
-      urlConn.setRequestProperty(HttpHeaders.AUTHORIZATION,
-                                 String.format("%s %s", authenticator.getType(), authenticator.getCredentials()));
+      setAuthHeader(urlConn::setRequestProperty, HttpHeaders.AUTHORIZATION, authenticator.getType(),
+                    authenticator.getCredentials());
     }
 
-    String internalCredentials = authenticationContext.getPrincipal().getCredential();
+    Principal principal = authenticationContext.getPrincipal();
```

Comment:
You still have this duplication. What.I mean is to create method, e.g.
```
void fillAuthentication(..., BiConsumer<String,String> headerSetter) {
  String userID = null;
    Credential internalCredentials = null;
    if (principal != null) {
      userID = principal.getName();
      internalCredentials = principal.getCredential();
    }
    if (internalCredentials != null) {
      headerSetter.apply(Authorization, String.format("%s %s", Constants.Security.Headers.RUNTIME_TOKEN,
                    internalCredentials.getType().getQualifiedName(), internalCredentials.getValue()));
    }
    if (userID != null) {
      headerSetter.apply(Constants.Security.Headers.USER_ID, userID);
    }
}
```
And then `fillAuthentication(..., urlConn::setRequestPropery)` and `fillAuthentication(..., headers::put)`. This way you get rid of a big code dup block

Reply:
Oh yeah let me extract that out, fixed